labgrid/driver/sshdriver.py

```diff
@@ -135,28 +135,41 @@ def _start_own_master_once(self, timeout):
+            version = re.match(r"^OpenSSH_(\d+)\.(\d+)", version.stderr)
+            major, minor = map(int, version.groups())
+
+            if major < 8 or (major == 8 and minor < 5):
+                ssh_good_version = False
+            else:
+                ssh_good_version = True
+
+            return_value = 0
             subprocess_timeout = timeout + 5
-            stdout, _ = self.process.communicate(timeout=subprocess_timeout)
-            if self.process.returncode != 0:
-                stdout = stdout.split(b"\n")
-                for line in stdout:
-                    self.logger.warning("ssh: %s", line.rstrip().decode(encoding="utf-8", errors="replace"))
-
-                try:
-                    with open(f'{self.tmpdir}/proxy-stderr') as proxy_err_fd:
-                        proxy_error = proxy_err_fd.read().strip()
-                    if proxy_error:
-                        raise ExecutionError(
-                            f"Failed to connect to {self.networkservice.address} with {' '.join(args)}: error from SSH ProxyCommand: {proxy_error}",  # pylint: disable=line-too-long
-                            stdout=stdout,
-                        )
-                except FileNotFoundError:
-                    pass
-
-                raise ExecutionError(
-                    f"Failed to connect to {self.networkservice.address} with {' '.join(args)}: return code {self.process.returncode}",  # pylint: disable=line-too-long
-                    stdout=stdout,
-                )
+            if not ssh_good_version:
+                return_value = self.process.wait(timeout=subprocess_timeout)
+            elif ssh_good_version or return_value != 0:
+                stdout, _ = self.process.communicate(timeout=subprocess_timeout)
+                if self.process.returncode != 0:
+                    stdout = stdout.split(b"\n")
+                    for line in stdout:
+                        self.logger.warning("ssh: %s", line.rstrip().decode(encoding="utf-8", errors="replace"))
+
+                    try:
+                        with open(f'{self.tmpdir}/proxy-stderr') as proxy_err_fd:
+                            proxy_error = proxy_err_fd.read().strip()
+                        if proxy_error:
+                            raise ExecutionError(
+                                f"Failed to connect to {self.networkservice.address} with {' '.join(args)}: error from SSH ProxyCommand: {proxy_error}",  # pylint: disable=line-too-long
+                                stdout=stdout,
+                            )
+                    except FileNotFoundError:
+                        pass
+
+                    raise ExecutionError(
+                        f"Failed to connect to {self.networkservice.address} with {' '.join(args)}: return code {self.process.returncode}",  # pylint: disable=line-too-long
+                        stdout=stdout,
+                    )
```

Comment:
Wouldn't it be sufficient if we call `communicate()` prior to `self.process.returncode != 0:` for OpenSSH >=8.5 and in the `self.process.returncode != 0:` block for OpenSSH <8.5? That would make the patch way shorter.

Reply:
Reasonable, see updated.